lib/namespace.rb

```diff
@@ -0,0 +1,80 @@
+require 'tmpdir'
+require 'fileutils'
+require 'rubygems'
+
+$MAIN = self
+
+module NameSpace
+  NAMESPACE_TMP_DIR = Dir.tmpdir
+
+  def self.define
+    ns = Module.new
+    ns.extend NameSpaceMethods
+    ns.instance_variable_set(:@features, [])
+    ns.instance_variable_set(:@realpaths, {}) # realpath => Qtrue
+    ns.instance_variable_set(:@realpath_map, {}) # feature => realpath
+    ns.instance_variable_set(:@ext_handles, {}) # name => handle
+    ns.instance_variable_set(:@LOAD_PATH, $LOAD_PATH.dup)
+    ns
+  end
+
+  module NameSpaceMethods
+    private def in_this_namespace
+      prev_ns = $CURRENT_NAMESPACE
+      $CURRENT_NAMESPACE = self
+      yield
+    ensure
+      $CURRENT_NAMESPACE = prev_ns
```

Comment:
This is not thread-safe, an easy fix would be to use Fiber-local variables instead (`Thread.current[:name]`)